runtime/compiler/net/RawTypeConvert.hpp

```diff
@@ -0,0 +1,290 @@
+         }
+      };
+
+   template <typename... T> struct RawTypeConvert<const std::tuple<T...>>
+      {
+      static inline std::tuple<T...> onRecv(Message::DataDescriptor *desc)
+         {
+         return TupleTypeConvert<0, T...>::onRecvImpl(static_cast<Message::DataDescriptor *>(desc->getDataStart()));
+         }
+
+      template <typename Tuple, size_t... Idx>
+      static inline uint32_t onSendImpl(Message &msg, const Tuple &val, index_tuple_raw<Idx...>)
+         {
+         uint32_t descIndex = msg.getNumDescriptors();
+         uint32_t descOffset = msg.reserveDescriptor();
+         uint32_t totalSize = TupleTypeConvert<0, T...>::onSendImpl(msg, std::get<Idx>(val)...);
+         // Only get pointer to the tuple descriptor after all of its elements were serialized,
+         // because buffer might have been reallocated
+         Message::DataDescriptor *desc = msg.getDescriptor(descIndex);
+         desc->type = Message::DataDescriptor::DataType::TUPLE;
+         desc->size = totalSize;
+         return totalSize;
+         }
+      
+      static inline uint32_t onSend(Message &msg, const std::tuple<T...> &val)
+         {
+         using Idx = typename index_tuple_gen_raw<sizeof...(T)>::type;
+         return onSendImpl(msg, val, Idx());
+         }
+      };
+
+   template <typename... T> struct RawTypeConvert<std::tuple<T...>> : RawTypeConvert<const std::tuple<T...>> { };
+
+   // setArgs fills out a message with values from a variadic argument list.
+   // calls RawTypeConvert::onSend for each argument.
+   template <typename Arg1, typename... Args>
+   struct SetArgsRaw
+      {
+      static void setArgs(Message &message, Arg1 &arg1, Args&... args)
+         {
+         SetArgsRaw<Arg1>::setArgs(message, arg1);
+         SetArgsRaw<Args...>::setArgs(message, args...);
+         }
+      };
+   template <typename Arg1>
+   struct SetArgsRaw<Arg1>
+      {
+      static void setArgs(Message &message, Arg1 &arg1)
+         {
+         RawTypeConvert<Arg1>::onSend(message, arg1);
+         }
+      };
+   template <typename... Args>
+   void setArgsRaw(Message &message, Args&... args)
+      {
+      message.getMetaData()->numDataPoints = sizeof...(args);
+      SetArgsRaw<Args...>::setArgs(message, args...);
+      }
+
+   template <typename Arg1, typename... Args>
+   struct GetArgsRaw
+      {
+      static std::tuple<Arg1, Args...> getArgs(const Message &message, size_t n)
+         {
+         return std::tuple_cat(GetArgsRaw<Arg1>::getArgs(message, n), GetArgsRaw<Args...>::getArgs(message, n + 1));
+         }
+      };
+   template <typename Arg>
+   struct GetArgsRaw<Arg>
+      {
+      static std::tuple<Arg> getArgs(const Message &message, size_t n)
+         {
+         Message::DataDescriptor *desc = message.getDescriptor(n);
+         // TODO: add type checking functionality
+         // if (data.type_case() != AnyPrimitive<typename ProtobufTypeConvert<Arg>::ProtoType>::typeCase())
```

Comment:
This type checking is important, but don't know how to do it.
Maybe use it above on all those onRecv functions because the template will tell us what we expect and the descriptor tells us what we got.

Reply:
Yeah, that would probably be the easiest thing to do.